exporters/elasticsearch/src/es_log_exporter.cc

```diff
@@ -0,0 +1,235 @@
+   */
+  void OnResponse(http_client::Response &response) noexcept override
+  {
+    // Store the body of the request
+    body_ = std::string(response.GetBody().begin(), response.GetBody().end());
+
+    // Set the response_received_ flag to true and notify any threads waiting on this result
+    response_received_ = true;
+    cv_.notify_all();
+  }
+  
+  /** 
+   * A method the user calls to block their thread until the response is received, or the timeout is exceeded.
+   */
+  bool waitForResponse(unsigned int timeoutSec = 1)
+  {
+    std::mutex mutex_;
+    std::unique_lock<std::mutex> lk(mutex_);
+    cv_.wait_for(lk, std::chrono::milliseconds(1000 * timeoutSec));
+    return response_received_;
+  }
+
+  /** 
+   * Returns the body of the response
+   */
+  std::string GetResponseBody()
+  {
+    if(!response_received_)
+      return "No response";
+    
+    return body_;
+  }
+
+  // Virtual method definition that isn't used
+  void OnEvent(http_client::SessionState state,
+                       opentelemetry::nostd::string_view reason) noexcept override
+  {}
+  
+
+private:
+  // Define a mutex and condition variable
+  std::condition_variable cv_;
+
+  // Whether the response from Elasticsearch has been received
+  bool response_received_ = false;
+
+  // A string to store the response body
+  std::string body_ = "";
+};
+
+ElasticsearchLogExporter::ElasticsearchLogExporter() : 
+  options_(ElasticsearchExporterOptions()) {}
+
+ElasticsearchLogExporter::ElasticsearchLogExporter(const ElasticsearchExporterOptions &options) : 
+  options_{options} 
+{
+  // Create a connection to the ElasticSearch instance
+  // TODO: Figure out why not working
+  // session_ = session_manager_.CreateSession(options_.host_, options_.port_);
+}
+
+sdklogs::ExportResult ElasticsearchLogExporter::Export(
+    const nostd::span<std::unique_ptr<opentelemetry::logs::LogRecord>> &records) noexcept
+{
+  // Return failure if this exporter has been shutdown
+  if (isShutdown_)
+  {
+    return sdklogs::ExportResult::kFailure;
+  }
+
+  // Create a json array to store all the JSON log records
+  json logs = json::array();
+
+  for (auto &record : records)
+  {
+    // Convert the log record to a JSON object 
+
+    logs.emplace_back(RecordToJSON(std::move(record)));
+  }
+
+  // Create a connection to the ElasticSearch instance
+  auto session = session_manager_.CreateSession(options_.host_, options_.port_);
+  auto request = session->CreateRequest();
+
+  // Populate the request with headers and methods
+  request->SetUri(options_.index_ + "/_bulk?pretty");
+  request->SetMethod(http_client::Method::Post);
+  request->AddHeader("Content-Type", "application/json");
+
+  // Add the request body
+  std::string body = "";
+  for(int i = 0; i < logs.size(); i++)
+  {
+    // Append {"index":{}} before JSON body, which tells Elasticsearch to write to index specified in URI
+    body += "{\"index\" : {}}\n";
```

Comment:
Should this be outside the for loop?

Reply:
Nope, its specific to the Bulk API where before each JSON object you send you need to tell it to store in the default index. Yea it's a strange format.

Reply:
Is it absolutely necessary though? In this example, there's no `index: {}` specified?
<img width="744" alt="Screen Shot 2020-12-11 at 6 21 54 PM" src="https://user-images.githubusercontent.com/25284332/101963840-c0991080-3bdd-11eb-84b6-6728f4c86417.png">

Reply:
In this section of the API: https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html#bulk-api-request-body 

It also state the `_index` field is optional?

Reply:
The example that I ran locally is 
```
curl -X POST 'http://localhost:9200/my-index-000001/_bulk?pretty' -H 'Content-Type: application/json' --data-binary '
{"index" : {}}
{"name": "Mark","age": 22}
{"index":{}}
{"name": "Karen","age": 21}
'
```
which works properly. If I remove the two `index:{}` lines, making the command 
```
curl -X POST 'http://localhost:9200/my-index-000001/_bulk?pretty' -H 'Content-Type: application/json' --data-binary '
{"name": "Mark","age": 22}
{"name": "Karen","age": 21}
'
```
then Elasticsearch throws the indexing error: 
```
"type" : "illegal_argument_exception",
 "reason" : "Malformed action/metadata line [2], expected START_OBJECT or END_OBJECT but found [VALUE_STRING]
```
I've tried many different ways of writing the request without including these index lines, but Elasticsearch seems to need it to work.

Reply:
It hink its because you're sending each field of the json object as a single request. Couldn't the attributes be combined into an json array? Like` [{"name": "Mark","age": 22}, {"name": "Karen","age": 21}]`

Reply:
Or `{"Mark": 22, "Karen": "21"}`

Reply:
Since it's using the same curl POST request, then all the JSON objects are definitely not being separated into separate HTTP requests. I found other people doing a similar approach online, so I think this way is the intended way of doing it. ([link1](https://stackoverflow.com/questions/61170157/newline-error-in-elasticsearch-bulk-api-post-request) and [link2](https://stackoverflow.com/questions/48579980/elasticsearch-bulk-request-throws-error-in-elasticsearch-6-1-1))